activemodel/lib/active_model/serializers/json.rb

```diff
@@ -146,9 +235,46 @@ def as_json(options = nil)
       def from_json(json, include_root = include_root_in_json)
         hash = ActiveSupport::JSON.decode(json)
         hash = hash.values.first if include_root
+        hash = json_key_formatters[:from_json].format_keys!(hash)
         self.attributes = hash
         self
       end
+
+      class KeyFormatter # :nodoc:
+        class Identity
+          def format_keys!(hash)
+            hash
+          end
+        end
+
+        def initialize(*args, **options)
+          @format = {}
+          @cache = {}
+
+          args.each do |name|
+            @format[name] = []
+          end
+          options.each do |name, parameters|
+            @format[name] = parameters
+          end
+        end
+
+        def format_keys!(hash)
+          hash.deep_transform_keys! do |key|
+            format(key)
+          end
+        end
+
+        def format(key)
+          @cache[key] ||= @format.inject(key.to_s) do |result, (func, args)|
+            if func in ::Proc
+              func.call(result, *args)
+            else
+              result.send(*func, *args)
+            end
```

Comment:
Since this draws inspiration from `jbuilder`, support for symbols and hash pairs is built-in. 

However, since the advent of anonymous proc arguments like `_1`, the kind of legibility-focused support proposed by Symbol and hash pairs might not be worthwhile. For example, compare the following:

```ruby
# proc with _1
key_format from_json: -> { _1.camelize :lower }

# hash pair
key_format from_json: { camelize: :lower }
```

Since the first is implemented with built-in Ruby and the second is clever and requires special treatment, it might be worth simplifying to callables-only.